crates/phactory/src/system/mod.rs

```diff
@@ -866,6 +815,81 @@ impl<Platform: pal::Platform> System<Platform> {
             info!("Gatekeeper: successfully decrypt received master key");
             self.set_master_key(master_pair, true);
         }
+        Ok(())
+    }
+
+    fn process_contract_key_distribution(
+        &mut self,
+        block: &mut BlockInfo,
+        origin: MessageOrigin,
+        event: DispatchContractKeyEvent<chain::Hash, chain::BlockNumber, chain::AccountId>,
+    ) -> Result<(), TransactionError> {
+        if !origin.is_gatekeeper() {
+            error!("Invalid origin {:?} sent a {:?}", origin, event);
+            return Err(TransactionError::BadOrigin);
+        }
+
+        // TODO(shelven): forget contract key after expiration time
+        let keypair = sr25519::Pair::restore_from_secret_key(&event.secret_key);
+        let contract_key = ContractKey(keypair);
+        let contract_pubkey = contract_key.public();
+        let contract_info = event.contract_info;
+        match contract_info.code_index {
+            CodeIndex::NativeCode(_contract_id) => {
+                // TODO(shelven): launch native code instance
+            }
+            CodeIndex::WasmCode(code_hash) => {
+                let code = chain_state::read_contract_code(block.storage, code_hash);
+                if let Some(code) = code {
```

Comment:
Readability: try to reduce the indent level by return early in branches.

Bad example:

```rust
if cond1 {
  if cond2 {
    if cond3 {
      return Ok(())
    }
    return Err(..);
  }
} else {
  return Err(..);
}
```

Good:

```rust
if !cond1 {
  return Err(..);
}
if !cond2 {
  return Err(..);
}
if !cond3 {
  return Err(..);
}
Ok(())
```

Reply:
done